src/server/common/helpers/auth/federated-oidc.js

```diff
@@ -0,0 +1,239 @@
+    scope: settings.scope,
+    code_challenge: codeChallenge,
+    code_challenge_method: 'S256'
+  }
+
+  if (!oidcConfig.serverMetadata().supportsPKCE()) {
+    logger.debug("server doesn't support PKCE")
+    nonce = openid.randomNonce()
+    parameters.nonce = nonce
+  }
+
+  request.yar.set(settings.sessionName, {
+    codeVerifier,
+    nonce
+  })
+
+  const refererPath = getRefererAsRelativeURL(request?.info?.referrer, '/')
+  request.yar.flash(sessionNames.referrer, refererPath)
+
+  return openid.buildAuthorizationUrl(oidcConfig, parameters)
+}
+
+/**
+ * Post-login, user has logged to AAD and has been redirected back to the auth callback.
+ * If the token is validated then the credentials are returned.
+ */
+async function postLogin(request, oidcConfig, settings) {
+  const state = request.yar.get(settings.sessionName, true)
+  const codeVerifier = state?.codeVerifier
+  const nonce = state?.nonce
+
+  // These values are set in the first part, if they're missing its probably because
+  // they've gone directly to the redirect link or refreshed it etc.
+  Hoek.assert(
+    codeVerifier || nonce,
+    'No verifier set in session, try logging in again.'
+  )
+
+  // `currentUrl` must match the full external url, including hostname.
+  const currentUrl = asExternalUrl(request.url, config.get('appBaseUrl'))
+
+  logger.info('validating token')
+  const token = await openid.authorizationCodeGrant(oidcConfig, currentUrl, {
+    pkceCodeVerifier: codeVerifier,
+    expectedNonce: nonce,
+    idTokenExpected: true
+  })
+
+  Hoek.assert(token, 'Failed to validate token')
+
+  const expiresIn = token.expiresIn()
+  const claims = token.claims()
+  return {
+    expiresIn,
+    token: token.access_token,
+    accessToken: token.access_token,
+    refreshToken: token.refresh_token,
+    idToken: token.id_token,
+    claims,
+    // This bit is fairly portal specific, maybe build it in the callback handler?
+    profile: {
+      id: claims.oid,
+      displayName: claims.name,
+      email: claims.email ?? claims.preferred_username
+    }
+  }
+}
+
+/**
+ * Refreshes the client credentials using a refresh token.
+ * Decorates the request object as `request.refreshToken(token)`
+ * @param {{}} settings
+ * @param {string} jwtRefreshToken
+ * @returns {Promise<{}>}
+ */
+async function refreshToken(settings, jwtRefreshToken) {
+  const federatedToken = await settings.tokenProvider()
+
+  const oidcConfig = await openid.discovery(
+    new URL(settings.discoveryUri),
+    settings.clientId,
+    {},
+    ClientFederatedCredential(federatedToken)
+  )
+
+  try {
+    return await openid.refreshTokenGrant(oidcConfig, jwtRefreshToken, {
+      scope: settings.scope
+    })
+  } catch (e) {
+    logClientError('refreshToken failed', e)
+    throw e
+  }
+}
+
+function logClientError(msg, err) {
+  if (err instanceof openid.ClientError) {
+    logger.error(
+      `${msg}: ${err.name} ${err.code} ${err.message}\n${err.cause.stack}`
```

Comment:
If you send the error object to the first argument in `logger.error` pino logs out all the parts. The object needs to be called `error`

```   
logger.error(error, error.message)
```